### DISCUSSION TITLE
Querying over entities that use single table inheritance (sti)

### DISCUSSION THREAD

**Discussion by yackinn** [mikro-orm/mikro-orm#5931] [General]
How to correctly query for inherited entities in this case?

```ts
@Entity({
  discriminatorColumn: 'type',
  abstract: true,
})
class Base {
   @PrimaryKey()
   id: string

   @Enum(Type)
   type: Type;    

   @ManyToOne(() => Base, { nullable: true })
   parent?: Base | null;
}

@Entity({
    discriminatorValue: Type.ChildA
})
class ChildA extends Base {
   @Enum(Type)
   override type: Type.ChildA;    

   @ManyToOne(() => ChildA, { nullable: true })
   override parent?: ChildA | null;
}

// using this query the parent property is null
const entityA = em().findOne(Base, {populate: ["parent"]})

// using this query the parent property is an instance of ChildA
const entityB = em().findOne(ChildA, {populate: ["parent"]})
```
I guess there's missing metadata for ChildA on "parent" when using "Base" as entity.
What's the correct way of setting the entities up or creating the queries?

Essentially, we have base entities that all differ by certain properties. When someone performs a search they always want to search across all types, that's why we want to store the entities in one table.

**Comment by B4nan** [comment]
What queries does it produce? Child properties are inlined to the base, I guess you are confusing the ORM when you override that property with a different type. Your `child.parent` will always add the type condition and it might be happening now when you query the base too because of that.

Also, what version are you on? I was fixing some bugs with a similar setup recently.

**Reply by yackinn** [reply]
Thank you for your response.

I'm on version 6.3.4.

```sql
SELECT
	"c0".*,
	"p1"."id" AS "p1__id",
	"p1"."kind" AS "p1__kind",
	"p1"."parent_id" AS "p1__parent_id",
	"p1"."type" AS "p1__type", ...
FROM
	"public"."classification" AS "c0"
	LEFT JOIN "public"."classification" AS "p1" ON "c0"."parent_id" = "p1"."id"
		AND "p1"."type" = 'ze'  ---- this is incorrect; p1.type is not just "ze"; in case of entity OPS the parent.type is "ops"
WHERE
	"c0"."code" = '8-40'
LIMIT 1
```
I guess that's what you meant by saying it's inlined to the base.

```ts
class Ops extends Base {
    type = Type.OPS
}

class Ze extends Base {
    type = Type.OPS
}

```

**Reply by yackinn** [reply]
is it possible to remove the `AND "p1"."type" = 'ze'` portion from the query?

I guess it's added as ZeEntity is the last child entity of Classification that is processed by the orm when setting up the metadata.

**Reply by B4nan** [reply]
Remove your extension of the parent property and it will most likely work this way, as I said, its caused by that.

**Reply by yackinn** [reply]
removing the 
```ts
@Index()
@ManyToOne(() => Ops, { nullable: true })
override parent?: Ops;
```
means the type isn't narrowed correctly for calling services.

How would you deal with that? Using generic in the base entity?

**Reply by B4nan** [reply]
The whole issue is overriding the type of your property, that affects metadata of the base entity too. You are not supposed to be changing types of properties in children, and this is effectively the same, since you change the target entity type.

~What version are you on?~ sorry, i am blind

**Reply by yackinn** [reply]
```sql
SELECT
	"o0".*,
	"p1"."id" AS "p1__id",
	"p1"."parent_id" AS "p1__parent_id",
	"p1"."code" AS "p1__code",
	"p1"."type" AS "p1__type"
FROM
	"public"."classification" AS "o0"
	LEFT JOIN "public"."classification" AS "p1" ON "o0"."parent_id" = "p1"."id"
WHERE
	"o0"."code" = '8-40'
	AND "o0"."type" = 'ops'
LIMIT 1
```
As you said removing the override removes the type constraint from the query correctly. So that works.
But the parent property in the instantiated entity is null though.

**Reply by B4nan** [reply]
Can you provide a complete repro?

**Reply by yackinn** [reply]
You're absolutely right. It happened due to overriding fields on a child class. Once I removed all overrides it worked.

For others that stumble upon a similar issue, that's what I did.

```ts
export abstract class Classification<
  TClassificationType extends ClassificationType = ClassificationType
> extends BaseEntity {
  @Index()
  @Enum(ClassificationType)
  type: TClassificationType;

export class Child extends Classification<ClassificationType.ChildA> {
...
}
```

@B4nan Thanks a lot.
